wolfssl/wolfcrypt/types.h

```diff
@@ -1138,8 +1138,9 @@ typedef struct w64wrapper {
         WC_ALGO_TYPE_SEED = 5,
         WC_ALGO_TYPE_HMAC = 6,
         WC_ALGO_TYPE_CMAC = 7,
+        WC_ALGO_TYPE_CERT = 8,
```

Comment:
I'm still struggling with the idea of have a crypto callback for getting a cert and algo type cert. All of these changes could be isolated to the TLS layer. Is there a reason not to have the `wolfSSL_CTX_use_certificate_ex` API call directly into `Pkcs11GetCert` and avoid changes to crypto layer? I'm curious what other engineers think like @bigbrett and @billphipps and @douzzer and @JacobBarthelmeh

Reply:
I am certainly open to suggestions, I agree that the new additions are very different than the existing design. Adding cert as an "algo" felt awkward but I didnt see a better way at the time.

Reply:
I agree with @dgarske that we need to avoid representing TLS-layer stuff in the wolfcrypt layer.  We have some backward stuff already (`wolfcrypt/src/asn.c` depends on `GetCA()`, `GetCAByName()`, `GetCAByKeyHash()`, `GetCAByAKID()`, `wolfSSL_X509_NAME_add_entry_by_NID()`, `wolfSSL_X509_NAME_free()`, `wolfSSL_X509_NAME_new_ex()`) and it's an awkward PITA -- pause to admire the stub implementations in `wolfcrypt/src/asn.c` gated on `defined(WOLFCRYPT_ONLY)`.  Causes additional PITA for the linuxkm in FIPS mode.

So, it seems to me, we really need to do better, drawing and observing the abstraction barrier here.